akka-http-caching/src/main/scala/akka/http/caching/RefreshingCache.scala

```diff
@@ -0,0 +1,130 @@
+   * and an optional expiration on unused entries
+   */
+  def apply[K, V](cachingSettings: scaladsl.CachingSettings): akka.http.caching.scaladsl.Cache[K, V] = {
+    val settings = cachingSettings.refreshingCacheSettings
+
+    require(settings.maxCapacity >= 0, "maxCapacity must not be negative")
+    require(settings.refreshAfterWrite.isFinite, "refreshAfterWrite must be finite")
+
+    if (settings.expireAfterWrite.isFinite) expiringRefreshingCache(settings.maxCapacity, settings.refreshAfterWrite, settings.expireAfterWrite)
+    else simpleRefreshingCache(settings.maxCapacity, settings.refreshAfterWrite)
+  }
+
+  /**
+   * Java API
+   * Creates a new [[akka.http.caching.RefreshingCache]] using configuration of the system,
+   * with an asychronous refresh interval and an optional expiration on unused entries
+   */
+  def create[K, V](system: ActorSystem): akka.http.caching.javadsl.Cache[K, V] =
+    apply(system)
+
+  /**
+   * Java API
+   * Creates a new [[akka.http.caching.RefreshingCache]], with an asychronous refresh interval
+   * and an optional expiration on unused entries
+   */
+  def create[K, V](settings: javadsl.CachingSettings): akka.http.caching.javadsl.Cache[K, V] =
+    apply(settings.asScala)
+
+  private def simpleRefreshingCache[K, V](maxCapacity: Int, refreshAfterWrite: Duration): RefreshingCache[K, V] = {
+    val store = Caffeine.newBuilder().asInstanceOf[Caffeine[K, V]]
+      .maximumSize(maxCapacity)
+      .refreshAfterWrite(refreshAfterWrite.asJava)
+      .buildAsync[K, V]
+    new RefreshingCache[K, V](store)
+  }
+
+  private def expiringRefreshingCache[K, V](maxCapacity: Long, refreshAfterWrite: Duration, expireAfterWrite: Duration): RefreshingCache[K, V] = {
+
+    def expire: Caffeine[K, V] => Caffeine[K, V] = { builder =>
+      if (expireAfterWrite.isFinite) builder.expireAfterWrite(expireAfterWrite.toMillis, TimeUnit.MILLISECONDS)
+      else builder
+    }
+
+    def refresh: Caffeine[K, V] => Caffeine[K, V] = { builder =>
+      if (refreshAfterWrite.isFinite) builder.refreshAfterWrite(refreshAfterWrite.toMillis, TimeUnit.MILLISECONDS)
+      else builder
+    }
+
+    val builder = Caffeine.newBuilder().asInstanceOf[Caffeine[K, V]]
+      .maximumSize(maxCapacity)
+
+    val store = (refresh andThen expire)(builder).buildAsync[K, V]
+    new RefreshingCache[K, V](store)
+  }
+
+  def toJavaMappingFunction[K, V](genValue: () => Future[V]): BiFunction[K, Executor, CompletableFuture[V]] =
+    asJavaBiFunction[K, Executor, CompletableFuture[V]]((k, e) => genValue().toJava.toCompletableFuture)
+
+  def toJavaMappingFunction[K, V](loadValue: K => Future[V]): BiFunction[K, Executor, CompletableFuture[V]] =
+    asJavaBiFunction[K, Executor, CompletableFuture[V]]((k, e) => loadValue(k).toJava.toCompletableFuture)
+}
+
+/** INTERNAL API */
+@InternalApi
+private[caching] class RefreshingCache[K, V](val store: AsyncLoadingCache[K, V]) extends Cache[K, V] {
```

Comment:
Is it even necessary to duplicate most of the implementation here? What kind of API changes are needed for this feature?

Reply:
I think the big thing is that AsyncLoadingCache will require the caller to pass a loading function in on construction otherwise it will have some dumb default:
```scala

    implicit val dummyLoader: AsyncCacheLoader[K, V] =
      (_: K, _: Executor) => Future.failed(new NoSuchElementException("")).toJava.toCompletableFuture
```

So it requires an AsyncCacheLoader to be passed in on construction to the store, if you're okay with that api change here?

Reply:
but yeah otherwise, I think the C can actually stay the same and just be common between the too.

so all of
```scala
private[caching] class LfuCache[K, V](val store: AsyncCache[K, V]) extends Cache[K, V] {
```
could probably just be
```scala
private[caching] class CacheImpl[K, V](val store: AsyncCache[K, V]) extends Cache[K, V] {
```

and then AsyncCacheLoader adheres to AsyncCache in the caffeine code other than a few default function changes